src/focus.service.ts

```diff
@@ -416,25 +416,35 @@ export class FocusService {
     const { width, height } = rect;
     let { top, left } = rect;
 
-    for (let parent = el.parentElement; parent !== null; parent = parent.parentElement) {
+    for (let parent = el.parentElement; parent !== container.parentElement; parent = parent.parentElement) {
 
       // Special case: treat the body as the viewport as far as scrolling goes.
-      const prect = parent === document.body
-        ? { top: 0, left: 0, height: window.innerHeight, width: window.innerWidth }
+      const prect = parent === container
+        ? {
+          top: Number(window.getComputedStyle(container, null).paddingTop.slice(0, -2)),
+          left: 0,
+          height: container.clientHeight - Number(window.getComputedStyle(container, null).paddingBottom.slice(0, -2)),
+          width: container.clientWidth
+        }
         : parent.getBoundingClientRect();
 
       // Trigger if this element has a vertical scrollbar
       if (parent.scrollHeight > parent.clientHeight) {
         const scrollTop = parent.scrollTop;
-        const showsBottom = scrollTop + (top - prect.top + height) - prect.height;
+        const showsBottom = scrollTop + top + (height - prect.height);
         const showsTop = scrollTop + (top - prect.top);
 
         if (showsTop < scrollTop) {
-          animate(parent, showsTop, scrollTop, x => parent.scrollTop = x);
-          top += scrollTop - showsTop;
+          animate(parent, showsTop, scrollTop, x => {
+            parent.scrollTop = x;
+            top += scrollTop - showsTop;
```

Comment:
Sorry isn't this what you want? Without updating top in the callback what I've found is that when animate needs to be called twice in a single scrolling event, the scrolling will happen as though only the second scrolling happened. And this is because top is updated before the animation has actually happened, therefore the second animate is calculated based on an inaccurate representation of the page